libcontainer/rootfs_linux.go

```diff
@@ -1079,8 +1088,37 @@ func pivotRoot(rootfs string) error {
 		return &os.PathError{Op: "fchdir", Path: "fd " + strconv.Itoa(newroot), Err: err}
 	}
 
-	if err := unix.PivotRoot(".", "."); err != nil {
-		return &os.PathError{Op: "pivot_root", Path: ".", Err: err}
+	pivotErr := unix.PivotRoot(".", ".")
+	if errors.Is(pivotErr, unix.EINVAL) {
+		// If pivot_root(2) failed with -EINVAL, one of the possible reasons is
+		// that we are in early boot and trying pivot_root on top of the
+		// initramfs (which isn't allowed because initramfs/rootfs doesn't have
+		// a parent mount).
+		//
+		// Traditionally, users were told to pass --no-pivot-root (which used a
+		// chroot instead) but this is very insecure (even with the hardenings
+		// we've put into our chroot() wrapper).
+		//
+		// A much better solution is to create a bind-mount of the target and
+		// chroot into it, resulting in a parented mount that pivot_root(2)
+		// will accept. One minor issue is that the mount will still exist (and
+		// in the case of an init system like systemd, this will result in
+		// wasted memory, so they have to do some hacks to clear the initramfs)
+		// but the mount is masked in a much more safe way than chroot() so
+		// this is still much better.
+		if err := unix.Mount(".", ".", "", unix.MS_BIND|unix.MS_REC, ""); err != nil {
```

Comment:
This is probably taking the same kernel locks the bindfd protection for the /proc/self/exe was taking, right?

Can we be very sure this path is only used by users that don't have a lot of churn? like if all kata containers need to do this, then we will run into the same problems we run with bindfd, I guess?

Reply:
This is being done in a mount namespace, we already do a lot of `mounts` in our mount namespace to configure the container.

Reply:
This is being done in a mount namespace, we already do a lot of `mounts` in our mount namespace to configure the container.

It should be noted that the mount seqlocks aren't per-namespace but the issue with `bindfd` was that the mounts were leaked into the host. While it is true that doing mounts is not cheap, I'm not entirely convinced that taking mount locks at all is a problem (despite what https://github.com/opencontainers/runc/pull/3599#discussion_r1228889317 suggested). We can't construct a container without doing mounts, which requires taking global VFS locks.

Reply:
Oh, makes sense. Although this will be used on exec too, not just on container creation. crun doesn't have such a flag, containerd doesn't seem to use it by default (and in v2 runtime it seems to not even be exposed), so that is nice. I'm open to try :)

Reply:
We don't do this for `runc exec` -- we join the container namespace which is already `pivot_root`ed.